File: sdk/python/trialmcp_client/fhir.py
"""FHIR sub-client for clinical data operations.

Wraps the four FHIR server tools: ``fhir_read``, ``fhir_search``,
``fhir_patient_lookup``, and ``fhir_study_status``.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

from .config import ServerName
from .exceptions import raise_for_error
from .models import FHIRResource, FHIRSearchResult, StudyStatusResult

if TYPE_CHECKING:
    from .client import TrialMCPClient

logger = logging.getLogger("trialmcp.fhir")

_SERVER = ServerName.FHIR


class FHIRClient:
    """Sub-client for the ``trialmcp-fhir`` MCP server.

    All returned patient data is de-identified by the server.  The SDK
    never stores or logs raw patient identifiers.

    Parameters
    ----------
    parent:
        The :class:`TrialMCPClient` that owns this sub-client.
    """

    def __init__(self, parent: TrialMCPClient) -> None:
        self._client = parent

    # ------------------------------------------------------------------
    # fhir_read
    # ------------------------------------------------------------------

    async def read(
        self,
        *,
        resource_type: str,
        resource_id: str,
        version: str | None = None,
    ) -> FHIRResource:
        """Read a single FHIR resource by type and ID.

        Parameters
        ----------
        resource_type:
            FHIR resource type (e.g. ``"Patient"``, ``"ResearchStudy"``).
        resource_id:
            The pseudonymised resource identifier.
        version:
            Optional specific version to retrieve.

        Returns
        -------
        FHIRResource:
            The de-identified FHIR resource.

        Raises
        ------
        NotFoundError:
            If the resource does not exist.
        InvalidInputError:
            If the resource type or ID is invalid.
        """
        params: dict[str, Any] = {
            "resource_type": resource_type,
            "resource_id": resource_id,
        }
        if version:
            params["version"] = version

        result = await self._client.call_tool(_SERVER, "fhir_read", params)
        raise_for_error(result)
        return FHIRResource.from_dict(result)

    # ------------------------------------------------------------------
    # fhir_search
    # ------------------------------------------------------------------

    async def search(
        self,
        *,
        resource_type: str,
        query_params: dict[str, str] | None = None,
        max_results: int = 100,
        page_token: str | None = None,
    ) -> FHIRSearchResult:
        """Search for FHIR resources matching the given criteria.

        Parameters
        ----------
        resource_type:
            FHIR resource type to search.
        query_params:
            Search parameters as key-value pairs.
        max_results:
            Maximum number of results to return.
        page_token:
            Continuation token for paginated results.

        Returns
        -------
        FHIRSearchResult:
            The matching resources and pagination metadata.
        """
        params: dict[str, Any] = {
            "resource_type": resource_type,
            "max_results": max_results,
        }
        if query_params:
            params["query_params"] = query_params
        if page_token:
            params["page_token"] = page_token

        result = await self._client.call_tool(_SERVER, "fhir_search", params)
        raise_for_error(result)
        return FHIRSearchResult.from_dict(result)

    # ------------------------------------------------------------------
    # fhir_patient_lookup
    # ------------------------------------------------------------------

    async def patient_lookup(
        self,
        *,
        patient_pseudonym: str,
        include_observations: bool = False,
        include_conditions: bool = False,
    ) -> FHIRResource:
        """Look up a de-identified patient by pseudonymised identifier.

        Parameters
        ----------
        patient_pseudonym:
            The HMAC-SHA256 pseudonymised patient identifier.
        include_observations:
            Whether to include clinical observations in the bundle.
        include_conditions:
            Whether to include diagnosed conditions.

        Returns
        -------
        FHIRResource:
            The de-identified patient resource (possibly a Bundle).
        """
        params: dict[str, Any] = {
            "patient_pseudonym": patient_pseudonym,
        }
        if include_observations:
            params["include_observations"] = True
        if include_conditions:
            params["include_conditions"] = True

        result = await self._client.call_tool(_SERVER, "fhir_patient_lookup", params)
        raise_for_error(result)
        return FHIRResource.from_dict(result)

    # ------------------------------------------------------------------
    # fhir_study_status
    # ------------------------------------------------------------------

    async def study_status(
        self,
        *,
        study_id: str,
    ) -> StudyStatusResult:
        """Retrieve the current status of a clinical study.

        Parameters
        ----------
        study_id:
            The clinical study identifier (e.g. NCT number).

        Returns
        -------
        StudyStatusResult:
            Current study status, enrolment, and site information.
        """
        result = await self._client.call_tool(
            _SERVER,
            "fhir_study_status",
            {"study_id": study_id},
        )
        raise_for_error(result)
        return StudyStatusResult.from_dict(result)

    # ------------------------------------------------------------------
    # Convenience helpers
    # ------------------------------------------------------------------

    async def search_all(
        self,
        *,
        resource_type: str,
        query_params: dict[str, str] | None = None,
        max_total: int = 1000,
    ) -> list[FHIRResource]:
        """Auto-paginate through all matching FHIR resources.

        Parameters
        ----------
        resource_type:
            FHIR resource type to search.
        query_params:
            Search parameters.
        max_total:
            Safety limit on total resources fetched.

        Returns
        -------
        list[FHIRResource]:
            All matching resources up to *max_total*.
        """
        all_resources: list[FHIRResource] = []
        page_token: str | None = None

        while len(all_resources) < max_total:
            batch_size = min(100, max_total - len(all_resources))
            result = await self.search(
                resource_type=resource_type,
                query_params=query_params,
                max_results=batch_size,
                page_token=page_token,
            )
            all_resources.extend(result.resources)

            if not result.next_page_token:
                break
            page_token = result.next_page_token

        return all_resources

    async def read_research_study(self, study_id: str) -> FHIRResource:
        """Convenience wrapper to read a ResearchStudy resource.

        Parameters
        ----------
        study_id:
            The study resource identifier.

        Returns
        -------
        FHIRResource:
            The ResearchStudy FHIR resource.
        """
        return await self.read(resource_type="ResearchStudy", resource_id=study_id)

    async def search_research_subjects(
        self,
        *,
        study_id: str,
        status: str | None = None,
    ) -> FHIRSearchResult:
        """Search for research subjects enrolled in a study.

        Parameters
        ----------
        study_id:
            The study identifier to filter by.
        status:
            Optional subject status filter.

        Returns
        -------
        FHIRSearchResult:
            Matching research subject resources.
        """
        query: dict[str, str] = {"study": study_id}
        if status:
            query["status"] = status
        return await self.search(
            resource_type="ResearchSubject",
            query_params=query,
        )

File: sdk/typescript/src/fhir.ts
/**
 * FHIR client for the National MCP-PAI Oncology Trials network.
 *
 * Wraps the trialmcp-fhir MCP server tools:
 *   - fhir_read: Read a FHIR resource by type and ID
 *   - fhir_search: Search FHIR resources with parameters
 *   - fhir_patient_lookup: Look up patient enrollment by pseudonym
 *   - fhir_study_status: Get clinical study status and enrollment metrics
 *
 * All patient identifiers use pseudonyms; real PHI never leaves the server.
 */

import { TrialMcpClient } from './client';
import { InvalidInputError, NotFoundError } from './errors';
import type {
  FhirResource,
  FhirBundle,
  FhirReadResult,
  FhirSearchParams,
  PatientLookupResult,
  StudyStatusResult,
} from './models';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const SERVER: 'trialmcp-fhir' = 'trialmcp-fhir';

/** FHIR resource types relevant to oncology trials */
const ONCOLOGY_RESOURCE_TYPES = [
  'Patient',
  'ResearchStudy',
  'ResearchSubject',
  'Observation',
  'Condition',
  'MedicationAdministration',
  'MedicationRequest',
  'Procedure',
  'DiagnosticReport',
  'ImagingStudy',
  'Consent',
  'Encounter',
  'CarePlan',
  'AdverseEvent',
  'ClinicalImpression',
] as const;

/** Maximum search results per query */
const MAX_SEARCH_RESULTS = 500;

/** Pattern for valid FHIR resource IDs */
const RESOURCE_ID_PATTERN = /^[A-Za-z0-9\-._]+$/;

// ---------------------------------------------------------------------------
// FHIR Client
// ---------------------------------------------------------------------------

export class FhirClient {
  private readonly client: TrialMcpClient;

  constructor(client: TrialMcpClient) {
    this.client = client;
  }

  /**
   * Read a single FHIR resource by type and ID.
   *
   * @param resourceType - FHIR resource type (e.g. 'Patient', 'ResearchStudy')
   * @param resourceId - Logical resource ID
   * @returns The FHIR resource with optional validation metadata
   * @throws NotFoundError if the resource does not exist
   * @throws InvalidInputError if resource type or ID is invalid
   */
  async read(resourceType: string, resourceId: string): Promise<FhirReadResult> {
    this.validateResourceType(resourceType);
    this.validateResourceId(resourceId);

    const result = await this.client.callTool<FhirReadResult>(SERVER, 'fhir_read', {
      resource_type: resourceType,
      resource_id: resourceId,
    });

    return result.data;
  }

  /**
   * Read a FHIR resource and return just the resource object.
   * Convenience wrapper around read() that unwraps the result.
   */
  async readResource(resourceType: string, resourceId: string): Promise<FhirResource> {
    const result = await this.read(resourceType, resourceId);
    return result.resource;
  }

  /**
   * Search FHIR resources with query parameters.
   *
   * Supports standard FHIR search parameters including modifiers.
   * Results are returned as a FHIR Bundle (searchset).
   *
   * @param resourceType - FHIR resource type to search
   * @param params - Search parameters as key-value pairs
   * @param maxResults - Maximum results to return (default: 100, max: 500)
   * @returns FHIR Bundle containing matching resources
   */
  async search(
    resourceType: string,
    params: Record<string, string> = {},
    maxResults: number = 100,
  ): Promise<FhirBundle> {
    this.validateResourceType(resourceType);

    if (maxResults <= 0 || maxResults > MAX_SEARCH_RESULTS) {
      throw new InvalidInputError(
        `maxResults must be between 1 and ${MAX_SEARCH_RESULTS}, got ${maxResults}`,
        { server: SERVER, tool: 'fhir_search', details: { field: 'maxResults' } },
      );
    }

    const result = await this.client.callTool<FhirBundle>(SERVER, 'fhir_search', {
      resource_type: resourceType,
      params,
      max_results: maxResults,
    });

    return result.data;
  }

  /**
   * Search for ResearchStudy resources by trial phase.
   * Convenience method for common oncology trial queries.
   */
  async searchStudiesByPhase(phase: string, status?: string): Promise<FhirBundle> {
    const params: Record<string, string> = { phase };
    if (status) params.status = status;
    return this.search('ResearchStudy', params);
  }

  /**
   * Search for adverse events associated with a study.
   */
  async searchAdverseEvents(studyId: string): Promise<FhirBundle> {
    return this.search('AdverseEvent', { study: studyId });
  }

  /**
   * Look up a patient's enrollment status using their pseudonym.
   *
   * This tool uses pseudonymous identifiers only; real patient
   * identifiers are never transmitted over the MCP network.
   *
   * @param pseudonym - Patient pseudonym assigned at enrollment
   * @param trialId - Clinical trial identifier
   * @returns Patient enrollment and consent status
   */
  async patientLookup(pseudonym: string, trialId: string): Promise<PatientLookupResult> {
    if (!pseudonym || pseudonym.trim().length === 0) {
      throw new InvalidInputError('Patient pseudonym is required', {
        server: SERVER,
        tool: 'fhir_patient_lookup',
        details: { field: 'pseudonym' },
      });
    }

    if (!trialId || trialId.trim().length === 0) {
      throw new InvalidInputError('Trial ID is required', {
        server: SERVER,
        tool: 'fhir_patient_lookup',
        details: { field: 'trial_id' },
      });
    }

    const result = await this.client.callTool<PatientLookupResult>(
      SERVER,
      'fhir_patient_lookup',
      { pseudonym, trial_id: trialId },
    );

    return result.data;
  }

  /**
   * Get the status and enrollment metrics for a clinical study.
   *
   * Returns aggregate study-level data including enrollment counts,
   * site counts, and study phase information.
   *
   * @param studyId - Clinical study identifier
   * @returns Study status with enrollment metrics
   */
  async studyStatus(studyId: string): Promise<StudyStatusResult> {
    if (!studyId || studyId.trim().length === 0) {
      throw new InvalidInputError('Study ID is required', {
        server: SERVER,
        tool: 'fhir_study_status',
        details: { field: 'study_id' },
      });
    }

    const result = await this.client.callTool<StudyStatusResult>(
      SERVER,
      'fhir_study_status',
      { study_id: studyId },
    );

    return result.data;
  }

  /**
   * Retrieve multiple resources by type and IDs in parallel.
   *
   * @param resourceType - FHIR resource type
   * @param resourceIds - Array of resource IDs to fetch
   * @returns Array of resources (nulls for not-found)
   */
  async readMany(
    resourceType: string,
    resourceIds: string[],
  ): Promise<(FhirResource | null)[]> {
    const results = await Promise.allSettled(
      resourceIds.map((id) => this.readResource(resourceType, id)),
    );

    return results.map((r) =>
      r.status === 'fulfilled' ? r.value : null,
    );
  }

  // -------------------------------------------------------------------------
  // Validation
  // -------------------------------------------------------------------------

  private validateResourceType(resourceType: string): void {
    if (!resourceType || !/^[A-Z][a-zA-Z]+$/.test(resourceType)) {
      throw new InvalidInputError(
        `Invalid FHIR resource type '${resourceType}'. Must be PascalCase (e.g. 'Patient')`,
        {
          server: SERVER,
          details: { field: 'resource_type', expectedPattern: '^[A-Z][a-zA-Z]+$' },
        },
      );
    }
  }

  private validateResourceId(resourceId: string): void {
    if (!resourceId || !RESOURCE_ID_PATTERN.test(resourceId)) {
      throw new InvalidInputError(
        `Invalid resource ID '${resourceId}'. Contains prohibited characters`,
        {
          server: SERVER,
          details: { field: 'resource_id', expectedPattern: '^[A-Za-z0-9\\-._]+$' },
        },
      );
    }
  }
}
